### PR TITLE
[DNM] Add clock select support for ptl

### PR DESCRIPTION
**Pull request by dnikodem**
Added the option to select the link clock source when creating the SSP module.

**Important**: changes in SOF repo are necessary here - [WiP](https://github.com/thesofproject/sof/pull/9325)

### CODE REVIEW

File: drivers/dai/intel/ssp/ssp.c [hunk 1]
```diff
@@ -1885,12 +1892,21 @@ static int dai_ssp_parse_aux_data(struct dai_intel_ssp *dp, const void *spec_con
...
-	pre_aux_len = sizeof(*blob) + blob->i2s_mclk_control.mdivrcnt * sizeof(uint32_t);
-	aux_len = cfg_len - pre_aux_len;
-	aux_ptr = (uint8_t *)blob + pre_aux_len;
+	if (blob15->version == SSP_BLOB_VER_1_5) {
+		cfg_len = blob15->size;
+		pre_aux_len = sizeof(*blob) + blob15->i2s_mclk_control.mdivrcnt * sizeof(uint32_t);
+		aux_len = cfg_len - pre_aux_len;
+		aux_ptr = (uint8_t *)blob + pre_aux_len;
+	} else {
+		cfg_len = spec_config_data->config_length * 4;
```
Comment (singalsu) [inline drivers/dai/intel/ssp/ssp.c:1904]: Would `* sizeof(uint32_t) ` be suitable here also?

Comment (dnikodem) [reply drivers/dai/intel/ssp/ssp.c:1904]: Yes, I will correct it.

**Review by singalsu — DISMISSED** [review]

File: drivers/dai/intel/ssp/dai-params-intel-ipc4.h [hunk 2]
```diff
@@ -70,6 +70,35 @@ enum dai_intel_ipc4_connector_node_id_type {
...
+	uint32_t dma_channel_id;	/**< channel ID if DMA pre-allocated by host, otherwise
+					  *   invalid channel ID
+					  */
+	uint32_t stream_id;
+	/* DmaStreamChannelMap */
+	struct dma_stream_channel_map channel_map;
+	uint32_t dma_priv_config_size;
+	uint8_t  dma_priv_config[0];
+} __packed __aligned(4);
+
```
Comment (plbossart) [inline drivers/dai/intel/ssp/dai-params-intel-ipc4.h:101]: this is already defined in src/include/sof/audio/ipc-config.h

Please don't add new definitions that are NOT SSP-specific anyways,

File: drivers/dai/intel/ssp/ssp.c [hunk 3]
```diff
@@ -873,7 +873,9 @@ static void dai_ssp_program_channel_map(struct dai_intel_ssp *dp,
 		sys_write16((pcmsycm & 0xffff), reg_add);
 	}
 #elif defined(CONFIG_SOC_INTEL_ACE30_PTL)
-	const struct dai_intel_ipc4_ssp_configuration_blob *blob = spec_config;
+	const struct ipc4_copier_gateway_cfg *spec_config_data = spec_config;
+	const struct dai_intel_ipc4_ssp_configuration_blob *blob =
+		     (void *)spec_config_data->config_data;
 	uint64_t time_slot_map = 0;
```
Comment (plbossart) [inline drivers/dai/intel/ssp/ssp.c:879]: this is PTL-specific, the commit title mentioned support for LNL.

File: drivers/dai/intel/ssp/dai-params-intel-ipc4.h (hunk 2 above)
Comment (plbossart) [inline drivers/dai/intel/ssp/dai-params-intel-ipc4.h:101]: And this addition seems irrelevant anyways, the DMA TLV is added after the SSP blob. Do you actually need this structure?

**Review by plbossart — CHANGES_REQUESTED** [review]

**Review by andyross — CHANGES_REQUESTED** [review]
No complaints about the hardware side (DTS changes and the register interface), but the other changes here seem like a pretty terrible layering violation.  What you want to do is define a simple interface for your clock control driver (doesn't need to be fancy, just an enumerant of clocking states and a set/get predicate to switch between them would work, I assume?), and have the SOF code call that based on all the protocol minutiae.

File: drivers/dai/intel/ssp/dai-params-intel-ipc4.h (hunk 2 above)
Comment (plbossart) [reply drivers/dai/intel/ssp/dai-params-intel-ipc4.h:101]: Humm, not sure why?

The way I understand things is that the gateway config is made of
a) a versioned blob (created by topology or retrieved from NHLT), followed by
b) the DMA TLV (added at run-time by the Linux host driver)

The versioned blob has a fixed size and presumably contains the MLCS information, so in theory you can get to it by just jumping the size of the versioned_blob, minus the MLCS information size.

Put differently, the DMA information is totally unrelated to the SSP blob, and it's really awkward to refer to it for something that's only in the scope of the SSP opaque blob.

**Review by marc-hb — COMMENTED** [review]
>  drivers: ssp: update program of MLCS register
> 
> Programming of the MLCS register was performed on the incorrect bits.
> Additionally, saving the new version did not erase the previously set
> value, which could result in an incorrect register value.

This is sneaking a generic bug fix by hiding under the addition of a platform-specific feature... :-(

The commit title for a bug fix should use the verb "fix..." instead of "update..." so anyone looking at the git log can identify it correctly.

Also, how bad was that bug? If bad enough then it should be 1. filed and described somewhere 2. submitted separately for validation purposes.

**Comment by dnikodem** [comment]
> > drivers: ssp: update program of MLCS register
> > Programming of the MLCS register was performed on the incorrect bits.
> > Additionally, saving the new version did not erase the previously set
> > value, which could result in an incorrect register value.
> 
> This is sneaking a generic bug fix by hiding under the addition of a platform-specific feature... :-(
> 
> The commit title for a bug fix should use the verb "fix..." instead of "update..." so anyone looking at the git log can identify it correctly.
> 
> Also, how bad was that bug? If bad enough then it should be 1. filed and described somewhere 2. submitted separately for validation purposes.

@marc-hb my intention was not to hide any bug fix - I found this during development clock select for PTL. In my opinion it is not critical cause probably no one will change the MLCS register value in runtime to select different clock.

I will create separate GH issue and PR with fix for this. Thanks
